core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -680,13 +865,43 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
 	fn validate(&self, context: &mut ValidatorContext<Block>, who: &PeerId, data: &[u8])
 		-> network_gossip::ValidationResult<Block::Hash>
 	{
-		let (action, broadcast_topics) = self.do_validate(who, data);
+		let (action, broadcast_topics, peer_reply) = self.do_validate(who, data);
+
+		{
+			let mut inner = self.inner.write();
+			if let Some(peer) = inner.peers.peer(who) {
+				if peer.view.set_id == inner.local_view.set_id &&
+					peer.view.round.0.saturating_sub(2) > inner.local_view.round.0
+				{
+					// send catch up request if none pending
+					let round = peer.view.round.0 - 1; // peer.view.round is > 0
+					let request = CatchUpRequestMessage {
+						set_id: peer.view.set_id,
+						round: Round(round),
+					};
+
+					if inner.note_catch_up_request(who, &request) {
+						trace!(target: "afg", "Sending catch-up request for round {} to {}",
+							round,
+							who,
+						);
+
+						let message = GossipMessage::<Block>::CatchUpRequest(request);
+						context.send_message(who, message.encode());
```

Comment:
You could assign an `Option<GossipMessage>` to the "result" of this entire scope, and then if there is a message to send, do the `context.send_message(who, message.encode());`(without holding the lock).